### PR TITLE
Move Buffer to an export with a global alias

### PR DESCRIPTION
**Pull request by sandersn**
Currently Buffer is declared as a global and then exported from inside the 'buffer' module. This is disallowed when there is a merge because its declarations start coming from a file outside buffer.d.ts

However, this prevents Buffer merges from specifying what class they extend because the name is now private. That means pouchdb-core needs to change.

This is a pretty simple code move, but the diff is gigantic. I'm going to see if ignoring whitespace helps; if not I'll try something else to make it easier to read.
Edit: This diff is now quite small with whitespace ignored.

Might fix #54260; I'm not sure how to test this since pouchdb-core was already working via dtslint-runner.

### CODE REVIEW

**Comment by KapitanOczywisty** [comment]
Seems like now `pouch-db` have only incorrect Buffer definitions. Maybe buffer should be moved to standalone package for such cases.

```txt
Interface 'Buffer' incorrectly extends interface 'BufferModuleInternal'.
  Types of property 'write' are incompatible.
    Type '(string: string, offset?: number | undefined, length?: number | undefined, encoding?: string | undefined) => number' is not assignable to type '{ (string: string, encoding?: BufferEncoding | undefined): number; (string: string, offset: number, encoding?: BufferEncoding | undefined): number; (string: string, offset: number, length: number, encoding?: BufferEncoding | undefined): number; }'.
      Types of parameters 'offset' and 'encoding' are incompatible.
        Type 'BufferEncoding | undefined' is not assignable to type 'number | undefined'.
          Type '"ascii"' is not assignable to type 'number | undefined'.
```

I've seen [one package](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/3ca462d6776ef057c1794b2ed9f26d6abc04c8cf/types/indy-sdk/index.d.ts#L8) referencing [`buffer`](https://www.npmjs.com/package/buffer) package for `Buffer` definitions, but from my testing it is failing with node either.

**Comment by SimonSchick** [comment]
The `buffer` module can't really be separated since the `buffer` standalone package already exists (which has caused a fair amount of trouble).

**Comment by SimonSchick** [comment]
That said, shouldn't the package be targetting the 'buffer' package? Iirc the use the buffer polyfil.

**Comment by KapitanOczywisty** [comment]
In my testing using `buffer` package with node wasn't working either, and this is also importing polyfil code to projects which seems iffy.

About standalone buffer could be named `node-buffer` or so to avoid *some* confusion.

Btw. Is it possible to run one test file with `node` typings and one without?

**Comment by sandersn** [comment]
> In my testing using buffer package with node wasn't working either, and this is also importing polyfil code to projects which seems iffy.

pouchdb-core currently defines its own global Buffer type which has to merge with node's global Buffer. Is that better? I don't have a way to tell.

> Btw. Is it possible to run one test file with node typings and one without?

It's all in the same compilation, so: no. Once node's globals are loaded, they are available everywhere in the compilation.

File: types/pouchdb-core/index.d.ts [hunk 1]
```diff
@@ -16,8 +16,10 @@ interface Blob {
     slice(start?: number, end?: number, contentType?: string): Blob;
 }
 
-interface Buffer extends Uint8Array {
-    write(string: string, offset?: number, length?: number, encoding?: string): number;
+interface Buffer {
```
Comment (KapitanOczywisty) [inline types/pouchdb-core/index.d.ts:19]: Can you bring back `extends Uint8Array`? 2 methods are missing to fix errors:
```ts
    reverse(): this;
    subarray(start?: number, end?: number): Buffer;
```

Also you can add these missing methods:
```ts
    writeBigInt64BE(value: bigint, offset?: number): number;
    writeBigInt64LE(value: bigint, offset?: number): number;
    writeBigUInt64BE(value: bigint, offset?: number): number;
    writeBigUInt64LE(value: bigint, offset?: number): number;
    readBigUInt64BE(offset?: number): bigint;
    readBigUInt64LE(offset?: number): bigint;
    readBigInt64BE(offset?: number): bigint;
    readBigInt64LE(offset?: number): bigint;
```

**Comment by SimonSchick** [comment]
Can we not remove pouch-db's Buffer definition and make consumers install the respective polyfil typings?
I feel like having to change integral typings in order to support narrow edge cases is not going to be sustainable in the long run.

**Comment by KapitanOczywisty** [comment]
@SimonSchick 
- At this point Buffer definition is staying.
- Polyfil typings are also failing without these changes. It's maybe better to fix how TypeScript merges global interfaces.
- As I understand in browser `Blob` is used instead, but typings are for both enviroments.

File: types/pouchdb-core/index.d.ts (hunk 1 above)
Comment (KapitanOczywisty) [reply types/pouchdb-core/index.d.ts:19]: @sandersn `extends Uint8Array` is still needed to have all `Buffer` methods :)

Comment (sandersn) [reply types/pouchdb-core/index.d.ts:19]: No, the problem is that all `extends` clauses must be identical and the original one is `interface Buffer extends BufferModuleInteral {`. Inlining them as you suggest is the only way to get those methods.

Comment (KapitanOczywisty) [reply types/pouchdb-core/index.d.ts:19]: No, as far I know `interfaces` are merged by comparing properties, not by extension chain.

I've tested this PR's last commit and it's working fine with `extends Uint8Array` and `node`.

**Comment by sandersn** [comment]
@SimonSchick 

> I feel like having to change integral typings in order to support narrow edge cases is not going to be sustainable in the long run.

True, but Buffer needs to be written in a mergeable way in the same way that the DOM types are. The current export structure prevents that. And this PR's structure makes a little more sense to me: the declarations are inside the module, then exported from the `global` block that's nested inside the module.

> Can we not remove pouch-db's Buffer definition and make consumers install the respective polyfil typings?

@KapitanOczywisty I interpreted your comment "At this point Buffer definition is staying" to refer to pouchdb's Buffer definition. Is that correct? Can you elaborate why it needs to stay?

**Comment by KapitanOczywisty** [comment]
> > Can we not remove pouch-db's Buffer definition and make consumers install the respective polyfil typings?
> 
> @KapitanOczywisty I interpreted your comment "At this point Buffer definition is staying" to refer to pouchdb's Buffer definition. Is that correct? Can you elaborate why it needs to stay?

I was writing about what this PR is doing, I think there was a bit of misunderstanding what Simon was writing. Disregard that.

File: types/pouchdb-core/index.d.ts (hunk 1 above)
Comment (KapitanOczywisty) [reply types/pouchdb-core/index.d.ts:19]: > Inlining them as you suggest is the only way to get those methods.

`reverse` and `subarray` are redefined by `Buffer`, thus was needed to fix incompatible definitions when using `Uint8Array`.

`BigInt` were just missing in `pouchdb` definition.

**Comment by SimonSchick** [comment]
@sandersn I assume it needs to be mergeable to support class decl merging and/or static additions?
Iirc class decl merging should be coming at some point, would that fix it and allow us eventually rollback these changes, I'm mostly asking as it introduces some churn/issues for my autogenerated nodejs doc efforts.

**Comment by sandersn** [comment]
Well, class declaration merging would work around the previous `class`-to-dual-`interface` change. This change would still be needed, because `Buffer` needs to both be exported and available globally.

Where and when did you hear about class declaration merging possibly happening?

**Pull request closed without merging**